test-suite/js/src/test/scala/org/scalajs/testsuite/library/ReflectTest.scala

```diff
@@ -187,6 +192,23 @@ class ReflectTest {
     }
   }
 
+  @Test def testInnerObjectWithEnableReflectiveInstantiation_issue_3228(): Unit = {
+    assertFalse(Reflect.lookupLoadableModuleClass(NameInnerObject).isDefined)
+
+    val optClassData = Reflect.lookupInstantiatableClass(NameInnerObject)
+    assertTrue(optClassData.isDefined)
+    val classData = optClassData.get
+
+    val optCtor = classData.getConstructor(
+        classOf[ReflectTest.ClassWithInnerObjectWithEnableReflectiveInstantiation])
+    assertTrue(optCtor.isDefined)
+
+    /* We do not actually instantiate that, because it would violate the
+     * uniqueness of the inner object wrt. its enclosing class, which is not
+     * well-defined at the Scala level.
+     */
```

Comment:
This seems like a major oversight in the design of reflective instantiation. Should we even allow to reflectively instantiate non-static things? I understand that Java reflection will totally allow you to do this, but that doesn't mean we should support it.

Reply:
The alternative would be not to register non-static module classes, and therefore test that both `lookupLoadableModuleClass` and `lookupInstantiatableClass` return `None`. Would you prefer that? Or should we emit a compile error (that might be seriously limiting in some cases).

Note that reflectively instantiating non-static *classes* is perfectly well-defined, so I think we should not disallow that.

Reply:
I would suggest to extend the reflection API to allow reflective access to non-static modules (essentially a forwarder to the module accessor).

Not sure if we should extend `LoadableModuleClass` or add another type.

Reply:
I'm not sure how I feel about that. It might be the "best" solution, but the benefit/cost is close to 0. Nobody would actually use that API, AFAICT, and implementing it is non-trivial (need to recover the enclosing instance, find the right getter for the `object`, etc.).

Reply:
In that case, let's just not register it (and hence return `None`). However, we should probably emit a warning, otherwise this might cause some hard to debug problems (think of pass a `Class` to something, it fails to load this thing, why???). Then again, warning means that we need a suppression mechanism which is painful in scalac.

Alternatively we might return a `LoadableModuleClass` where `loadClass` throws an exception. This would also future proof the API in case we want to add "proper" nested module loading later.

Reply:
I don't think returning something for which `loadModule` throws an exception is more future-proof than returning `None`. The API would have to change anyway, because a nested-module-enabled `loadModule` would have to take the enclosing instance as argument.

I'd vote for returning `None`, and not even warn (I would warn if scalac had a mechanism to suppress warnings). Note that the API has *loadable* in its name (`lookupLoadableModuleClass`). We can argue that a nested module is not *loadable* per se, and therefore it is not found by this method call. This is similar to the fact that abstract classes are not found by `lookupInstantiatableClass`, because they're not instantiatable. We should document the behavior more clearly in the Scaladoc, though, for sure.

Reply:
Hmmm... fair enough. Abstract classes are in exactly the same boat, so let's go with returning `None` for now.

However, I really think we should improve the failure behavior of this API so that we are able to "fail differently" if an item is not found or because we can't return it due to "reasons". Showing these reasons will make it much easier for users to fix the problem at hand.

Think of: "Oh this is failing, it can't find the class... Did I get the name wrong (separators, trailing `$`, etc.)? Do I not inherit from the right thing? Wrong annotation?". All this debugging time can be avoided if we simply say: "A.B.C is abstract, can't instantiate" or "A.B.C is a non-static module, can't be loaded (because: ...)". 

Now, how to expose this in the API such that:
- One can still query if something is loadable without having to catch shady exceptions.
- One can reliably assume that if something gets returned it actually is loadable/instantiatable.

I do not know :)